libnetwork/drivers/bridge/setup_ip_tables_linux.go

```diff
@@ -31,11 +31,6 @@ const (
 )
 
 func setupIPChains(config configuration, version iptables.IPVersion) (natChain *iptables.ChainInfo, filterChain *iptables.ChainInfo, isolationChain1 *iptables.ChainInfo, isolationChain2 *iptables.ChainInfo, retErr error) {
-	// Sanity check.
-	if !config.EnableIPTables {
```

Comment:
I need to check if this is the "correct" fix here (but it's a bit confusing I guess);

- the `--iptables` flag was added to disable the daemon from creating iptables rules, so when it was added, it meant "don't make any changes to iptables")
    - ^^ disclaimer: IIRC there were still paths where it would still make changes (I _think_ the `--iptables` options applied to local networks, but not to overlay networks; unfortunately not all of that was well-documented by the people working on that)
- the `--ip6tables` option was added as a new feature


So if we unconditionally remove this check, then (at least for _this code_) we may be changing the meaning of the `--iptables=false` option (and it would now only disable iptables for IPv4)


So perhaps we need to have an (internal) `DisableIPTables` option that looks of the different combinations of `--iptables` and `--ip6tables`, so that _only_ setting `--iptables=false` will still disable all iptables code.

(do we need a `--ip4tables` option? or perhaps some new flag that allows configuring bits more granular?)

Reply:
> * the `--iptables` flag was added to disable the daemon from creating iptables rules, so when it was added, it meant "don't make any changes to iptables")

My understanding is that the `--iptables` flag is named after the [`iptables` utility](https://www.netfilter.org/projects/iptables/) it executes, which is IPv4-only.  The `--ip6tables` flag is named after the `ip6tables` utility, which is IPv6-only.  "Netfilter" is the name of the kernel-side logic that the `iptables` and `ip6tables` utilities manipulate.  If the `--iptables` flag was named `--netfilter` then I would expect it to cover both IPv4 and IPv6, such that `--netfilter=false` shuts off all packet filtering manipulations, both IPv4 and IPv6.

But it's not named `--netfilter`, and the existing code aligns with the separation based on IP version, except for this check.  This is the only place where both `config.EnableIPTables` *and* `config.EnableIP6Tables` must be simultaneously set.

> So if we unconditionally remove this check, then (at least for _this code_) we may be changing the meaning of the `--iptables=false` option (and it would now only disable iptables for IPv4)

This is a redundant check (the "sanity check" comment is accurate): All callers already ensure that either `config.EnableIPTables` (for IPv4) or `config.EnableIP6Tables` (for IPv6) is set.  So deleting this check does not change the behavior of either the `--iptables` flag or the `--ip6tables` flag (except it is now possible to pass `--iptables=false --ip6tables=true`).

> 
> So perhaps we need to have an (internal) `DisableIPTables` option that looks of the different combinations of `--iptables` and `--ip6tables`, so that _only_ setting `--iptables=false` will still disable all iptables code.
> 
> (do we need a `--ip4tables` option? or perhaps some new flag that allows configuring bits more granular?)

I think that setting the semantics of `--iptables` to cover both IPv4 and IPv6 would result in some confused users, due to the name similarity with the IPv4-only `iptables` utility.  (Why pick a flag name that's obviously based on that utility but give it different semantics?)

Of course `--iptables` is ambiguous to those unfamiliar with the `iptables` utility, and iptables is being replaced with [nftables](https://www.netfilter.org/projects/nftables/), so ideally we would replace these flags with something else.  My preference would be network- and container-level options, not daemon-level options.

Reply:
> My understanding is that the --iptables flag is named after the [iptables utility it executes, which is IPv4-only

No, the fact that it shells out to the `iptables` utility was an implementation detail, and because it was the only "portable" API; other ways to programmatically do the same because complicated _fast_  (plus at the time, docker already shelled out to most things, e.g. shelling out to the `lxc` command-line to configure things etc).

This flag was it was to "not manipulate iptables" (https://github.com/dotcloud/docker/issues/858); this was _waaaaaaay_ before custom networks even existed;
originally this was addressed by adding `-N none` (`--network=none`); https://github.com/moby/moby/pull/1260, but at the time it was already being discussed to have an explicit option to disable `iptables` configuration; which was added in docker 0.6;
https://github.com/moby/moby/blob/v0.6.5/docker/docker.go#L42

Flag description was improved a bit in docker 0.7;
https://github.com/moby/moby/blob/v0.7.0/docker/docker.go#L38

Reply:
I believe those pre-date IPv6 support, so "disable IPv4 rules" and "disable all rules" were synonymous back then. Whatever the original intention was, it was superseded when the `--ip6tables` flag was added (#41622).  From that PR, it looks like they intended `--iptables` to be limited to IPv4, not both IPv4 and IPv6 (though it's not perfectly clear):

> Add a config option `ip6tables` (like `iptables` for IPv4)

Reply:
I generally agree with you _IF_ we would working from a greenfield state, but we need to be careful here to change meaning of flags; at least in cases where we're _aware_ we're changing behavior (for sure, there's been cases where behavior was _unintentionally_ changed; if we discover those, and are still able to fix those, we should).

The tricky bit with this one is that `--iptables=false` (while it may still have some "rough edges") can be considered by users to be a security-setting to prevent the Docker Daemon from opening ports (or other changes in `iptables`). If setting another flag (`--ip6tables=true`) can silently step over that restriction, then that's a potential security issue (for sure, these settings require access to the daemon's configuration). Truth is that with over 30 million monthly installs (from our docker's package repository alone), it's really hard to tell all possible scenarios, but we should definitely not make those changes "unnoticed".

Perhaps we need new flags on the daemon to replace the existing ones that are less ambiguous (and possibly fix some of the limitations with the existing ones).

Reply:
moving conversation to https://github.com/moby/moby/issues/46467#issuecomment-1720586937